release-24.3: sql: Ensure we auto commit for first DDL in txn

spilchen opened the pull request:
Backport 1/1 commits from #141145.

/cc @cockroachdb/release

---

Previously, when the autocommit_before_ddl session setting was enabled, the first DDL statement in a multi-statement transaction did not trigger an auto-commit, leading to the use of the legacy schema changer. This behavior was observed in transactions such as:

```
SET autocommit_before_ddl = ON;
BEGIN;
DROP TABLE t1;
DROP TABLE t2;
```

This change will now ensure that we auto commit before the first drop. This ensures that we'll use the DSC for both drops.

Epic: None
Release note: None

Closes #141034
Release justification: we want to have consistent behaviour of autocommit_before_ddl in older releases. Mainly a test change.

spilchen committed:
sql: Ensure we auto commit for first DDL in txn

Previously, when the autocommit_before_ddl session setting was enabled,
the first DDL statement in a multi-statement transaction did not trigger
an auto-commit, leading to the use of the legacy schema changer. This
behavior was observed in transactions such as:

```
SET autocommit_before_ddl = ON;
BEGIN;
DROP TABLE t1;
DROP TABLE t2;
```

This change will now ensure that we auto commit before the first drop.
This ensures that we'll use the DSC for both drops.

Epic: None
Release note: None

Closes #141034

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141987)
<!-- Reviewable:end -->

rafiss commented:
I want to make sure we debug these testing failures before going ahead with the backport: https://github.com/cockroachdb/cockroach/issues/141878, https://github.com/cockroachdb/cockroach/issues/141884, https://github.com/cockroachdb/cockroach/issues/141903, https://github.com/cockroachdb/cockroach/issues/141899, https://github.com/cockroachdb/cockroach/issues/141881. I don't know if it's related yet, but seems like it could be

spilchen committed:
sql: Auto-commit before DDL in BIND

Several parts of conn_executor_*go already auto-commit transactions
before handling DDL commands. However, we were missing the auto-commit
for the BIND command. Without it, binding a prepared statement could
trigger the auto-commit after the bind, which closes the portal and
leads to unknown portal errors.

Epic: none
Release note (bug fix): Fixed a bug when running with
autocommit_before_ddl that could cause a runtime error when binding a
previously prepared DDL statement.

Fixes #141576
Fixes #141881

spilchen committed:
roachtest: Update lib/pq block list

The recent autocommit_before_ddl change (#141145) has affected three
lib/pq tests that were not designed with autocommit behavior in mind.

These tests were executing the following sequence:
```
BEGIN
CREATE TEMP TABLE
COPY
```

Since CREATE TEMP TABLE now triggers an auto-commit, the subsequent COPY
command fails because it must run inside an active transaction. The
error observed was: "pq: COPY is only allowed inside a transaction."

Epic: none
Release note: none

Closes #141899

spilchen committed:
roachtest: Update pgx block list

The pgx test v5.TestBeginReadOnly previously expected an error when
executing:

```
BEGIN READ ONLY
CREATE TABLE foo(id serial primary key)
```

Due to the recent autocommit_before_ddl change in #141145, the
transaction is auto-committed before the CREATE TABLE, allowing the
command to succeed. This update adjusts the block list so that we
ignore this failed test.

Epic: none
Release note: none

Closes #141903

spilchen committed:
roachtest: Update psycopg block list

This allows failures for some tests in the pscyopg tests that were due
to the recent autocommit_before_ddl change in #141145. Rationale for
each failing test are included as comments in the block list.

Epic: none
Release note: none
Closes #141884

spilchen committed:
roachtest: Update asyncpg block list

This allows failures for some tests in the asyncpg tests that were due
to the recent autocommit_before_ddl change in #141145. Rationale for
each failing test are included as comments in the block list.

Epic: none
Release note: none
Closes #141878

spilchen committed:
roachtest: Update knex expected test failures

This allows a failure for a test in the knex test suite that were due to
the recent autocommit_before_ddl change in #141145. Rationale for the
failing test is included as comment in the test driver.

Epic: none
Release note: none
Closes #141902
